sig-api-machinery KEP: Vanilla OpenAPI subset: structural schema

sttts opened the pull request:
This is the conversion of https://docs.google.com/document/d/1pcGlbmw-2Y0JJs9hsYnSBXamgG9TfWtHY6eh80zSTd8/edit?ts=5cbe0089#heading=h.l1xazlj9aw9f into the KEP format.

sttts commented:
/assign @liggitt

sttts commented:
/assign @apelisse @DirectXMan12 @mbohlool

sttts commented on keps/sig-api-machinery/20190425-structural-openapi.md:
```diff
@@ -0,0 +1,750 @@
+ 
+1. **structurality:** that both `ForbiddenGenerics` and `ForbiddenExtensions` only have zero values, with the two exceptions of (4).
+2. **completeness:** that the schema is _structurally complete_
+3. **RawExtension:** that for every schema with `x-kubernetes-embedded-resource: true` `x-kubernetes-preserve-unknown-fields: true` and `type: object` are set
```
@liggitt do we want to enforce `x-kubernetes-preserve-unknown-fields: true` and `type: object` as written here? I tend to do this as it restricts the special case even more.

liggitt replied:
sure

liggitt commented on keps/sig-api-machinery/20190425-structural-openapi.md:
```diff
@@ -0,0 +1,750 @@
+* Simplify the semantics of CRD validation OpenAPI v3 schemas to make implementation of advanced feature like pruning, defaulting, server-side apply and read-only validation considerably less difficult.
+* Support embedding of existing API types into a CRD via RawExtensions or via Golang type reuse (e.g. of PodSpecs).
+* Allow a rich OpenAPI v3 validation schema for CustomResource validation, but allow us to programmatically derive the restricted spec for all other purposes.
+* Perspective goal: have a canonical mapping from a restricted schema to Golang types and Protobuf v3.
```
s/Perspective/Prospective/?

sttts replied:
if that's a known term, yes. Some goal to align with future direction.

liggitt commented on keps/sig-api-machinery/20190425-structural-openapi.md:
```diff
@@ -0,0 +1,750 @@
+  
+Note: we explicitly accept that polymorphic types (like IntOrString) are an anti-pattern for Kubernetes-like types and we do not want to encourage its use beyond IntOrString.
+
+As a soft criterion to validate the direction, we want that restricted CRD OpenAPI schemas represent a model that can also be represented with other parts of the system: 
```
~~we want that restricted~~ CRD OpenAPI schemas *should*

liggitt commented on keps/sig-api-machinery/20190425-structural-openapi.md:
```diff
@@ -0,0 +1,750 @@
+
+**Structural Completeness**: all types and the possible object fields within a JSON object must be apparent from a structural schema without understanding the logical junctors `anyOf`, `allOf`, `oneOf` and `not`.
+
+Example:
```
make it clearer up front this is an example of what we want to avoid

sttts replied:
done

liggitt commented on keps/sig-api-machinery/20190425-structural-openapi.md:
```diff
@@ -0,0 +1,750 @@
+        // decoding step from pruning fields which are not specified 
+        // in the validation schema. This affects fields recursively,
+        // but switches back to normal pruning behaviour if nested 
+        // properties are specified in the schema.
```
> This affects fields recursively, but switches back to normal pruning behaviour if nested properties are specified in the schema.

Is there a clearer way to say this? Maybe something like:
> `x-kubernetes-preserve-unknown-fields: true` preserves unknown fields for this object (and any children of those unknown fields)

Thinking through this again, isn't that what `additionalProperties: true` means? "There may be unknown properties, some of which may be objects with arbitrary child fields"?

sttts replied:
Clarified during offline Slack discussion.

liggitt commented on keps/sig-api-machinery/20190425-structural-openapi.md:
```diff
@@ -0,0 +1,750 @@
+        XEmbeddedResource bool
+        
+        // x-kubernetes-int-or-string specifies that this value is 
+        // either an integer or a string. It is allowed to 
```
`It is allowed to ...`?

sttts replied:
fixed

liggitt commented on keps/sig-api-machinery/20190425-structural-openapi.md:
````diff
@@ -0,0 +1,750 @@
+     - minimum: 52
+       maximum: 60
+   ```
+   The later is what is also published.
````
typo: latter

liggitt commented on keps/sig-api-machinery/20190425-structural-openapi.md:
```diff
@@ -0,0 +1,750 @@
+    * rejected at CRD validation time on create and on update which sets the webhook; user gets validation error
+    * pre-existing webhooks on update will stay allowed
+  * publishing
+    * not published at publish time, `NonStructuralSchema condition is added to CRD status with a warning
```
```
`NonStructuralSchema`
```

liggitt commented on keps/sig-api-machinery/20190425-structural-openapi.md:
```diff
@@ -0,0 +1,750 @@
+    * pre-existing webhooks on update will stay allowed
+  * publishing
+    * not published at publish time, `NonStructuralSchema condition is added to CRD status with a warning
+` * pruning (which will be with explicit opt-in in `v1beta1`)
```
drop leading `

liggitt commented on keps/sig-api-machinery/20190425-structural-openapi.md:
````diff
@@ -0,0 +1,750 @@
+       x-kubernetes-preserve-unknown-fields: true
+       nullable: true
+   ```
+   and published the same way. Dropping value validation does not change anything.
````
can you clarify whether the "published" examples here refer to openapi v2 or v3? things like `nullable` don't exist in v2, right?

sttts replied:
clarified by giving both OpenAPI v3 and v2 output

liggitt commented on keps/sig-api-machinery/20190425-structural-openapi.md:
```diff
@@ -0,0 +1,750 @@
+**For Kubernetes 1.15, must haves:**
+
+* Implement structural-schema API validation for 1.15, but don't reject non-structural objects, only set the NonStructuralSchema condition.
+* Do not publish a CRD schema if NonStructuralSchema condition is true (pre-requisite of promoting publishing to beta).
```
even some of the structural schemas cannot be published in openapi v2, due to kubectl handling of additionalProperties, right?

sttts replied:
as we enforce types and properties outside of the logical junctors, we enforce that more of the spec is publishing (junctors are dropped today). So yes, not all goes through, but much more than before.

liggitt commented on keps/sig-api-machinery/20190425-structural-openapi.md:
```diff
@@ -0,0 +1,750 @@
+
+**For Kubernetes 1.15, stretch goals:**
+
+* Do not accept webhook conversion to be set if NonStructuralSchema condition is false.
```
nit: we wouldn't base validation on the condition, but on whether validation determined the CRD being submitted had only structural schemata

liggitt commented on keps/sig-api-machinery/20190425-structural-openapi.md:
```diff
@@ -0,0 +1,750 @@
+**For Kubernetes 1.15, stretch goals:**
+
+* Do not accept webhook conversion to be set if NonStructuralSchema condition is false.
+1.15, potentially as alpha if time permits:
```
if we intend to bring conversion to beta, whatever validation rejection we want must be in place

sttts replied:
added a note that webhook conversion will not be blocked by this, but we indent to make it conditional.

apelisse commented:
I don't think it's very clear from looking at the doc what is going to be published for openapi v2? My understanding is that the structural schema is going to be much easier to transform into this openapi v2. Also, I do expect the extensions to be persisted across conversion?

sttts commented:
@apelisse I clarified that the example show OpenAPI v3 publishing output. I also added the v2 variant.

I also added this note to clarify:

```markdown
Note: the OpenAPI `v2` output given above is the same we get from the [OpenAPI `v2` filtering in Kubernetes 1.14](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapi/conversion.go#L30): that filtering in the CRD OpenAPI `v2` publishing pipeline supports the unrestricted range of the CRD OpenAPI `v3 language and turns it into a weaker variant understandable by kubectl 1.13 and 1.14. Hence, our changes above have no influence on the OpenAPI `v2` publishing other than passing through of our new `x-kubernetes-*` vendor extensions.

Also note that structural schemas enforce types and properties outside of logical junctors. OpenAPI `v2` publishing will preserve these and hence, structural schemas lead to a more complete OpenAPI `v2` spec, client-side validation and `kubectl explain` output.
```

sttts commented on keps/sig-api-machinery/20190425-structural-openapi.md:
```diff
@@ -0,0 +1,750 @@
+  * IPv4 vs IPv6 strings: same fields with different formats
+  * embedded PodSpec
+  
+* discriminator union types (as described in [KEP: Union Types](https://github.com/kubernetes/enhancements/pull/926)).
```
fixed

sttts commented on keps/sig-api-machinery/20190425-structural-openapi.md:
```diff
@@ -0,0 +1,750 @@
+	// x-kubernetes-union-field-values lists the discriminator values 
+	// for mutual exclusive field names. This can only be set if
+	// x-kubernetes-union-discriminator is non-empty.
+        XUnionFieldValues      map[string]string
```
did you review an old version? I changed this to match @apelisse's proposal and implementation https://github.com/kubernetes-sigs/structured-merge-diff/pull/80. I don't propose anything new here around unions, but want to have them as an example.

sttts commented on keps/sig-api-machinery/20190425-structural-openapi.md:
```diff
@@ -0,0 +1,750 @@
+	Type        string
+	AdditionalProperties *StructuralSchemaOrBool
+	
+	Description string
```
sure. I thought title was forbidden. We can handle it the very same way as description.

sttts replied:
added.

sttts commented on keps/sig-api-machinery/20190425-structural-openapi.md:
```diff
@@ -0,0 +1,750 @@
+  
+  Potentially, we will also unfold `x-kubernetes-union-discriminator` to a schema restricting the discriminator's possibly values (using `pattern` or `enum`).
+
+These unfoldings are subject to change and explicitly not part of the API, but merely illustration and a rough sketch here. Code generators and other tools must not depend on the unfolding to happen in one way or another. It is guaranteed though that the `x-kubernetes-*` fields are preserved during publishing and hence can be used by tooling as an abstract hint for semantic complex to express in  ...
```
I meant e.g. client generators. They should not pattern match OpenAPI specs to find a union, but rather use the abstract `x-kubernetes-unions` tag. In other words: we express the logic in vanilla OpenAPI, but have abstract tags in addition to understand where complex logical OpenAPI expression came from.

sttts replied:
have reformulated this, ptal.

sttts commented:
@DirectXMan12 addressed your comments.

liggitt commented on keps/sig-api-machinery/20190425-structural-openapi.md:
```diff
@@ -0,0 +1,812 @@
+  Examples:
+  * string maps like labels and annotations, but also `map[string]int64`
+  * IPv4 vs IPv6 strings: same fields with different formats
+  * embedded PodSpec
```
does an embedded pod spec fit in this category? it has discriminated union types within it (like volumesource), as well as single and multi-key arrays (containers, ports, etc)

sttts replied:
Does it have any type we cannot express at all? (there are certain things we cannot validate completely, but the base types should be structural.

liggitt commented on keps/sig-api-machinery/20190425-structural-openapi.md:
````diff
@@ -0,0 +1,812 @@
+```
+In this example, `.foo` is specified as string even without looking at the `anyOf` construct, but `.bar`'s type is unclear without deeper semantic knowledge of `anyOf` and of minimum and maximum.
+
+We want that specified types and object field names are mechanically extractable with an understanding of `properties`, `items`, `additionaProperties` and `type` only. Other constructs like logical junctors should be usable to express complex validations, but should not be essential for the specification of the type structure of JSON values.
````
would this be clearer as:

> We want to be able to mechanically extract specified object field names and types using only the information specified in `properties`, `items`, `additionalProperties` and `type`.

sttts replied:
:+1:

liggitt commented on keps/sig-api-machinery/20190425-structural-openapi.md:
```diff
@@ -0,0 +1,812 @@
+	
+        // x-kubernetes-preserve-unknown-fields stops the API server 
+        // decoding step from pruning fields which are not specified 
+        // in the validation schema. This affects fields recursively,
```
am I reading too much into this comment, or are you distinguishing between StructuralSchema and ValidationSchema here?

sttts replied:
I don't. I have removed the word "validation". Less confusion that way.

liggitt commented on keps/sig-api-machinery/20190425-structural-openapi.md:
```diff
@@ -0,0 +1,812 @@
+        //      - type: string
+        //    - <NestedValueValidation>
+        ForbiddenGenerics Generic `"json:inline"`
+        ForbiddenExtensions Extensions `"json:inline"`
```
can we drop the json annotations to make it clearer that this isn't used for serialization yet?

liggitt commented on keps/sig-api-machinery/20190425-structural-openapi.md:
```diff
@@ -0,0 +1,812 @@
+1. `additionalProperties`, `description`, `title`, `nullable`, `default`, `readOnly` and `type` (with the two `x-kubernetes-int-or-string: true` exceptions) and the new `x-kubernetes-*` extensions cannot be used inside of logical junctors `allOf`, `anyOf`, `oneOf`, `not`
+2. When setting `StructuralSchema.ValueValidatoin` recursively to nil in a schema, the schema becomes weaker.
+
+The second consequence means that we can correctly validate an object without understanding value validations (i.e. without rejecting an object as invalid which is actually valid). This is crucial for knowingly incomplete schema algorithms as used in pruning, defaulting, apply and elsewhere.
```
"we can correctly validate *the existence and type of* an object's fields"?

sttts replied:
fixed

liggitt commented on keps/sig-api-machinery/20190425-structural-openapi.md:
```diff
@@ -0,0 +1,812 @@
+The additional validations have the following consequences:
+
+1. `additionalProperties`, `description`, `title`, `nullable`, `default`, `readOnly` and `type` (with the two `x-kubernetes-int-or-string: true` exceptions) and the new `x-kubernetes-*` extensions cannot be used inside of logical junctors `allOf`, `anyOf`, `oneOf`, `not`
+2. When setting `StructuralSchema.ValueValidatoin` recursively to nil in a schema, the schema becomes weaker.
```
typo: ValueValidation

liggitt commented on keps/sig-api-machinery/20190425-structural-openapi.md:
````diff
@@ -0,0 +1,812 @@
+    - type: string
+  - <ValueValidation.AnyOf value>
+  ```
+* `x-kubernetes-embedded-resource: true` unfolds to
````
apiVersion and kind would be required, correct?

sttts replied:
oh right. adding.

liggitt commented on keps/sig-api-machinery/20190425-structural-openapi.md:
```diff
@@ -0,0 +1,812 @@
+
+These unfoldings are subject to change and explicitly not part of the API, but merely illustration and a rough sketch here. Code generators and other tools must not depend on the unfolding to happen in one way or another. It is guaranteed though that the `x-kubernetes-*` fields are preserved during publishing and hence can be used by tooling as an abstract hint for semantics which are complex to  ...
+
+The guiding principle is that we don't want OpenAPI consuming tooling, e.g. client generators, to implement OpenAPI schema pattern heuristics to recover the original types like unions, RawExtensions or `IntOrString`. We support them with our abstract, officially supported vendor extensions for those cases.
```
nit: OpenAPI-consuming

liggitt commented on keps/sig-api-machinery/20190425-structural-openapi.md:
```diff
@@ -0,0 +1,812 @@
+        // restrict the embedded object. Both ObjectMeta and TypeMeta
+        // are validated automatically. x-kubernetes-preserve-unknown-fields
+        // must be true.
+        XEmbeddedResource bool
```
do we want to consider https://github.com/kubernetes/kubernetes/issues/76965 here, to allow recording one or more GVKs expected to be embedded here?

sttts replied:
I tend to postpone this whole `$ref` topic to later and to another KEP. Do you see that we have to decide anything now? We can add a `x-kubernetes-embedded-resource-gvk` is we want later.

liggitt replied:
sounds good

liggitt commented on keps/sig-api-machinery/20190425-structural-openapi.md:
```diff
@@ -0,0 +1,812 @@
+   type Foo struct {
+     // +nullable
+     // +kubebuilder:properties=apiVersion:pattern=v1
+     X runtime.RawExtensions
```
nit: `RawExtension`

liggitt commented on keps/sig-api-machinery/20190425-structural-openapi.md:
```diff
@@ -0,0 +1,812 @@
+
+Note: the OpenAPI v2 output given above is the same we get from the [OpenAPI v2 filtering in Kubernetes 1.14](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapi/conversion.go#L30): that filtering in the CRD OpenAPI v2 publishing pipeline supports the unrestricted range of the CRD OpenAPI v3 language and turns it into a weaker var ...
+
+Also note that structural schemas enforce types and properties outside of logical junctors. OpenAPI v2 publishing will preserve these and hence, structural schemas lead to a more complete OpenAPI v2 spec, client-side validation and `kubectl explain` output.
```
+1

sttts replied:
@roycaihw ^^ is a great improvement for your publishing code.

liggitt commented on keps/sig-api-machinery/20190425-structural-openapi.md:
```diff
@@ -0,0 +1,812 @@
+
+### Scope of necessary Changes
+
+* To implement this KEP we won't have to modify CRD validation in any way. We can keep using the go-openapi based validation algorithm.
```
we don't have to modify **custom resource** validation

liggitt commented on keps/sig-api-machinery/20190425-structural-openapi.md:
```diff
@@ -0,0 +1,812 @@
+
+* To implement this KEP we won't have to modify CRD validation in any way. We can keep using the go-openapi based validation algorithm.
+* To implement this KEP we won't have to modify kubectl in any way. We can though add support for the new `x-kubernetes-*` fields, though we will probably wait for OpenAPI v3 publishing and support in kubectl before doing that.
+* We don't have to change the API types as described above in the proposal, but we can implement the structural schema check via a number of additional validations on-top of the existing API types.
```
clarify these are CRD validations

liggitt commented on keps/sig-api-machinery/20190425-structural-openapi.md:
```diff
@@ -0,0 +1,812 @@
+
+* When creating `v1` CRDs, only structural schemas (with value-validation) are allowed.
+Updating a schema in a `v1` CRD to a non-structural schema is only allowed if the original schema was non-structural already.
+* Non-structural CRDs only allow `v1` schema updates if a structural schema is set.
```
is **already** set in the existing persisted object.

liggitt commented on keps/sig-api-machinery/20190425-structural-openapi.md:
```diff
@@ -0,0 +1,812 @@
+
+* Do not accept webhook conversion to be set if the schemas are not structural.
+
+  Note: we indent to bring webhook conversion to beta in 1.15. Hence, we won't block webhook conversion on this item, but indent to make it conditional on structural schemas if we are confident enough about the structural schema validation.
```
typos: intend

liggitt commented on keps/sig-api-machinery/20190425-structural-openapi.md:
```diff
@@ -0,0 +1,812 @@
+
+On the API surface we add
+* a `NonStructuralSchema` condition. This is binary compatible during upgrade and downgrade, i.e. no special migration steps are necessary for the cluster admin.
+* a `x-kubernetes-*` vendor extensions to the CRD type. These will be dropped on write on downgrade, which is to be expected and accepted for new fields.
```
expected and accepted for new fields **in beta resources**

liggitt commented:
looks great overall. a few nits and clarifications. only one substantial question at https://github.com/kubernetes/enhancements/pull/1002#discussion_r279429600

liggitt commented:
lgtm, can squash when other reviewers are happy

DirectXMan12 commented:
/lgtm

sttts commented:
Squashed

sttts commented:
@DirectXMan12 @liggitt please reapply the tag after squashing.

DirectXMan12 commented:
/lgtm

sttts commented:
@lavalamp @deads2k please take a look for approval. @liggitt and @DirectXMan12 lgtm'ed.

sttts commented:
/assign @deads2k @lavalamp

deads2k commented:
Glad to see this moving forward.

/approve

DirectXMan12 commented on keps/sig-api-machinery/20190425-structural-openapi.md:
```diff
@@ -0,0 +1,743 @@
+        // restrict the embedded object. Both ObjectMeta and TypeMeta
+        // are validated automatically. x-kubernetes-preserve-unknown-fields
+        // must be true.
+        XEmbeddedResource bool
```
It occurs to me at this point: why is this not a "format" value?  Same with "intorstring"

sttts replied:
format has local influence on the type it is defined in (are there exceptions?). `XEmbeddedResource` causes further behaviour changes which are not local (metadata is checked and pruned; apiVersion, kind is required). Also I think this is a very kubernetes specific concept. Using `format` for that doesn't feel right IMO, but is rather a cosmetical detail.

lavalamp commented:
I strongly approve of the concept and don't have time to nitpick any details, so LGTM!
